sys-devel/llvm/llvm16-16.0.6.recipe

```diff
@@ -0,0 +1,637 @@
+	llvm6${secondaryArchSuffix}_clang_analysis
+	llvm7${secondaryArchSuffix}_clang_analysis
+	llvm8${secondaryArchSuffix}_clang_analysis
+	llvm9${secondaryArchSuffix}_clang_analysis
+	llvm12${secondaryArchSuffix}_clang_analysis
+	llvm13${secondaryArchSuffix}_clang_analysis
+	llvm14${secondaryArchSuffix}_clang_analysis
+	llvm15${secondaryArchSuffix}_clang_analysis
+	"
+
+PROVIDES_lld="
+	llvm16${secondaryArchSuffix}_lld = $portVersion
+	cmd:ld.lld = $portVersion
+	cmd:ld64.lld = $portVersion
+	cmd:lld = $portVersion
+	cmd:lld_link = $portVersion
+	cmd:wasm_ld = $portVersion
+	devel:liblldCOFF$secondaryArchSuffix = $portVersion
+	devel:liblldCommon$secondaryArchSuffix = $portVersion
+	devel:liblldELF$secondaryArchSuffix = $portVersion
+	devel:liblldMinGW$secondaryArchSuffix = $portVersion
+	devel:liblldWasm$secondaryArchSuffix = $portVersion
+	"
+REQUIRES_lld="
+	haiku$secondaryArchSuffix
+	lib:libLLVM_16$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	"
+
+PROVIDES_libs="
+	llvm16${secondaryArchSuffix}_libs = $portVersion
+	lib:libclang_cpp$secondaryArchSuffix = $portVersionCompat
+	lib:libLLVM_16$secondaryArchSuffix = $portVersionCompat
+	lib:libLTO$secondaryArchSuffix = $portVersionCompat
+	lib:libRemarks$secondaryArchSuffix = $portVersionCompat
+	"
+REQUIRES_libs="
+	haiku$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	"
+
+PYTHON3_VERSION="3.9"
+PROVIDES_python3="
+	llvm16${secondaryArchSuffix}_python3 = $portVersion
+	cmd:lit
+	"
+REQUIRES_python3="
+	setuptools_python39
+	cmd:python$PYTHON3_VERSION
+	"
+
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libz$secondaryArchSuffix
+	setuptools_python39
+	"
+BUILD_PREREQUIRES="
+	cmd:cmake
+	cmd:diff
+	cmd:find
+	cmd:gcc$secondaryArchSuffix
+	cmd:grep
+	cmd:groff
+	cmd:ld$secondaryArchSuffix
+	cmd:make
+	cmd:ninja
+	cmd:python$PYTHON3_VERSION
+	cmd:sed
+	"
+
+BUILD()
+{
+	local cmakeFlags
+	if [ -n "$secondaryArchSuffix" ]; then
+		cmakeFlags=-DHAIKU_HYBRID_SECONDARY="\"${effectiveTargetArchitecture}\""
+		cmakeFlags="$cmakeFlags -DLLVM_LIBDIR_SUFFIX=/${effectiveTargetArchitecture}"
+	fi
+
+	# Haiku C++ requires rtti in a lot of central system components
+	# such as Mesa
+
+	# LLVM in default targets  AArch64, AMDGPU, ARM, AVR, BPF, Hexagon, Lanai, \
+	# Mips, MSP430, NVPTX, PowerPC, RISCV, Sparc, SystemZ, WebAssembly, X86, \
+	# XCore on Haiku. There is currently no good reason to build all of those.
+
+	if [ "$targetArchitecture" = x86_gcc2 ];then
+		export LIBRARY_PATH=$LIBRARY_PATH:$sourceDir/build/lib/x86
+	else
+		export LIBRARY_PATH=$LIBRARY_PATH:$sourceDir/build/lib
+	fi
+
+	cmake -S llvm -B build -G Ninja \
+		$cmakeDirArgs \
+		-DCMAKE_BUILD_TYPE=Release \
+		$cmakeFlags \
+		-DCMAKE_SKIP_RPATH=YES \
+		-DCOMPILER_RT_BUILD_SANITIZERS=OFF \
+		-DLLVM_ENABLE_RTTI=ON -DLLVM_LINK_LLVM_DYLIB=YES \
+		-DLIBUNWIND_ENABLE_STATIC=OFF \
+		-DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld;compiler-rt" \
+		-DLLVM_ENABLE_RUNTIMES="libunwind" \
```

Comment:
`libunwind` is enabled, but it is not included in this recipe.

It was included in the `llvm12` recipe as `llvm12_libunwind`. Any reason not to include it here?

Reply:
In 15 that I worked on, it was no longer a separate library, I think it became internal to llvm runtime libraries? You'll need to build to figure out where they've been relocated to.

Reply:
By that I mean libunwind moved from "projects" to "runtime" in configure step. It's no longer a "project".

Reply:
Hmm, that probably explains why dotnet has llvm-libunwind in its own tree. We probably have to upstream our libunwind patches to llvm then...

Reply:
I just took whatever is there in the llvm15 recipe, not sure if any change is needed?

Reply:
> By that I mean libunwind moved from "projects" to "runtime" in configure step. It's no longer a "project".

https://github.com/haikuports/haikuports/pull/7036/files

This is how I built `llvm-libgcc`, a "runtime" that is composed of `libunwind` and `compiler-rt` linked together into one single library.

I suppose `libunwind` can be compiled the same way, I am testing locally on top of the latest LLVM.

Reply:
> You'll need to build to figure out where they've been relocated to.

I've actually built the libraries, `libunwind.so` is located at `build/lib/libunwind.so`, just like all other LLVM libraries, if you enable `libunwind` in `LLVM_ENABLE_RUNTIMES`.

`libunwind` headers are located at `libunwind/include` just like older versions.

Reply:
`libunwind` can therefore be copied along with all other LLVM libraries (but I haven't seen this recipe doing such a thing) or be a separate package like before.

Reply:
In the current recipe `libunwind.so*` lands in `lib/$effectiveTargetMachineTriple/` and gets packed up in the `llvm16` base package along with the libunwind headers. I guess we can split them out to an `llvm16_libunwind` resp. `llvm16_libunwind_devel` package. What do you think, whould we move `libunwind.so*` to `lib/` or can they stay in the current place in `lib/$effectiveTargetMachineTriple/` ?

Reply:
It would make sense to split `libunwind` as a separate library since it is functionally complete and sometimes used as a standalone library dependent from LLVM. I therefore prefer `llvm16_libunwind` as a separate package.

However, if HaikuPorts authorities prefer to keep the package ecosystem simple, shipping it along with `llvm16` won't break anything (except probably some scripts using `"llvm"$version"_libunwind"` to customize the LLVM version).

Reply:
I've updated the recipe to build the libunwind packages